### ISSUE TITLE
@extend should unify parent selectors that share a simple selector

### ISSUE THREAD

**Issue by nategreen** [sass/sass#2570]
Hi! Sorry to (maybe) resurrect a long-dead issue, but I'm having the same problem as #1428, #1052 and others. I am attempting to do the same thing as the OP of #1052: create a system where **styles** are abstracted from **selectors**, so we can:

- use the same styles everywhere
- style differently-structured across projects—markup we can't necessarily change

Placeholders for the styles—`@extend`ed in a layer of selectors—was working well until I ran into the issue with how it combines multiple descendant selectors. (For example: [output I like](https://gist.github.com/nategreen/bfdb7852b0279c3a6a477d961c1029b8) \ [output I don't like](https://gist.github.com/nategreen/022fc0d0f0bc219284069b45f4e68cf0) ) I've tried using descendants in the placeholder layer and using them in the selector layer, but the result is the same: at some point, I (usually inadvertently) make it so the relationship between the placeholders doesn't match with the selectors', and I start to get selectors like `.thing.state .thing.other-state .part` when I need `.thing.state.other-state .part`.

I already know that I can work around this by creating a flat structure of placeholders/mixins with every permutation of states/styles encoded into their names, but I'm reluctant to settle for that—especially when my setup with placeholders for each style/state _almost_ works. (I'd really like to avoid forcing my teammates to remember/reproduce long names of states, parts, and/or variants in the correct order, e.g. 
 `%button-small-focus-active-background`.)

The team's responses to other issues in this vein make me think it'd be fruitless to ask for a change to placeholders or `@extend`, since our mental models about what those things _should_ be don't match up. Maybe some kind of Sass-specific combinator could solve the problem, or a flag on `@extend` that makes it...dumber?

Hopefully at the very least, you can see what I'm looking to solve for. Let me know if I need to clarify. Thanks!

**Comment by nex3** [comment]
You're actually running into a different problem than the issues you cite. Those issues are about Sass generating *too many selectors*, which it has to do because it's unaware of how the HTML that's being styled is structured. Your issue is that it's *not generating enough* selectors; specifically, it's not generating `.thing.state.other-state .part`.

The reason for this in general is that Sass trims selectors it judges unlikely to match real-world HTML, in order to avoid bloating the CSS and annoying even more people than those who filed the issues you linked. For example, the most theoretically-safe way to expand the selectors for

```scss
.a %b {x: y}
.c .d {@extend %b}
```

would be

```css
.a .c .d, .c .a .d, .a.c .d {
  x: y;
}
```

but Sass assumes that you *probably* aren't going to be putting `class="a c"` on the same element, and uses that to generate less CSS. (Users like those who filed #1428 and #1052 wish it would go even further and only generate one of `.a .c .d` and `.c .a .d`, but there's no way for it to know which one to choose.)

In your specific case, though, `.a` is `.custom-text-field:focus-within` and `.b` is `.custom-text-field.invalid`. These two compound selectors share the simple selector `.custom-text-field`, which is a good reason to believe that they'll apply to the same element, so Sass's heuristic to delete the unified selector fails.

I think this is worth tweaking the extension algorithm for. Specifically, I think we should generate the unified parent selector if and only if the two selectors being unified share at least one simple selector. For backwards compatibility, it should be generated in addition to the un-unified versions.

**Comment by nategreen** [comment]
Yes, thank you for articulating that better than I could. I think that after trying so many configurations I lost track of what I was trying to ask for. 👍

**Comment by nategreen** [comment]
So what's the process from here, then—how does this make its way through to dart-sass, about how long does it usually take, etc.? (Not to apply any undue pressure; just trying to get a sense of timing of this vs. my project.)

**Comment by nex3** [comment]
Since this is a language change, it needs to go through the language process. This means writing [a proposal](https://github.com/sass/language/tree/master/proposal) for the new behavior (in this case, probably an explanation of how to change [the existing `@extend` spec](https://github.com/sass/language/blob/master/spec/extend.md)), soliciting comments on that proposal (I usually just send out a tweet from [@SassCSS](https://twitter.com/SassCSS)), waiting a week or so for anyone interested to comment, *then* beginning the implementation.

Since this is a relatively bite-sized change, I'd *like* to get to it some time this quarter, but I'm not sure when I'll find the time exactly. If you wanted to take a crack at writing up the proposal, that would help a lot.

**Comment by nategreen** [comment]
Thanks! I don't know if I'll get a chance to submit a proposal before you do, but I will take a stab at it as soon as I can, if I don't see any activity. 👍

**Comment by nategreen** [comment]
Sorry for the long delay before circling back on this, haven't had time to try and dig into editing the spec but we also don't need/want this as urgently as 6 months ago; ran into performance/filesize problems after all, so our approach has changed and doesn't rely on `@extend` anymore.
